IMN-624 - Implementing Redis rate limiter

ecamellini opened the pull request:
Closes [IMN-624](https://pagopa.atlassian.net/browse/IMN-624)

[IMN-624]: https://pagopa.atlassian.net/browse/IMN-624?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Tested on local BFF

To test it, I configured it as follows (2 max requests every 10 seconds) and sent 3 requests in 10 seconds:
<img width="293" alt="image" src="https://github.com/pagopa/interop-be-monorepo/assets/6418684/08f9125c-7853-4c4f-98ae-d62282a120f6">

### Request 1
<img width="797" alt="Screenshot 2024-07-10 at 16 48 57" src="https://github.com/pagopa/interop-be-monorepo/assets/6418684/a2fbb050-eb0f-41e3-8096-432bc79de75e">

### Request 2
<img width="792" alt="Screenshot 2024-07-10 at 16 49 00" src="https://github.com/pagopa/interop-be-monorepo/assets/6418684/289b30bc-9aab-4d1e-b4ce-d3ff59debb4b">

### Requests 3 and 4
The burst rate limiter kicks in, and the burst percentage is 1.0 so two more requests are allowed (same amount).

<img width="792" alt="Screenshot 2024-07-10 at 16 49 00" src="https://github.com/pagopa/interop-be-monorepo/assets/6418684/289b30bc-9aab-4d1e-b4ce-d3ff59debb4b">

### Request 5
<img width="783" alt="Screenshot 2024-07-10 at 16 49 05" src="https://github.com/pagopa/interop-be-monorepo/assets/6418684/5155a126-a043-462e-8455-962e118047b5">
<img width="782" alt="Screenshot 2024-07-10 at 16 49 08" src="https://github.com/pagopa/interop-be-monorepo/assets/6418684/e3aca4d9-72b1-4437-89bc-29fa4147684b">

# Testing Redis timeout

Setting Redis timeout to 0 milliseconds:

<img width="746" alt="image" src="https://github.com/pagopa/interop-be-monorepo/assets/6418684/bbfbd930-0c7f-45c3-ab32-8a4f428a8db2">

Request passes:
<img width="795" alt="image" src="https://github.com/pagopa/interop-be-monorepo/assets/6418684/1b03addb-71f9-4cc7-a574-76a233034bb2">

🪄 Apparently, the rate limiter still has some internal logic to limit somehow the requests, because if I send many fast requests it still gives me the 429 error.

ecamellini commented on docker/docker-compose.yml:
```diff
@@ -133,9 +133,13 @@ services:
   # Mailpit is a local SMTP server, to replace AWS SES in local development
   mailpit:
-    image: axllent/mailpit
+    image: axllent/mailpit:v1.19
```
Unrelated, but since I am changing this file I am pinning also mailpit to a specific version

ecamellini replied:
You're right, I was just lazy 👼  Moved here: https://github.com/pagopa/interop-be-monorepo/pull/756

ecamellini commented on packages/commons-test/src/containerTestUtils.ts:
```diff
@@ -17,7 +17,7 @@ export const TEST_MINIO_IMAGE =
 export const TEST_MAILPIT_HTTP_PORT = 8025;
 export const TEST_MAILPIT_SMTP_PORT = 465;
-export const TEST_MAILPIT_IMAGE = "axllent/mailpit";
+export const TEST_MAILPIT_IMAGE = "axllent/mailpit:v1.19";
```
Unrelated, for consistency with https://github.com/pagopa/interop-be-monorepo/pull/725/files#r1672464296

AsterITA commented on packages/commons/src/rate-limiter/rateLimiterUtils.ts:
```diff
@@ -0,0 +1,15 @@
+  "X-RateLimit-Interval": rateLimiterStatus.rateInterval,
+  "X-RateLimit-Remaining": rateLimiterStatus.remainingRequests,
+});
+// TODO why are these different from the ones in the openapi spec?
```
IMHO: these are two different conventions, I would align the code to the openapi spec if possible

AsterITA reviewed (approved):
Let's wait @galales opinion too on this

galales commented on packages/commons/src/rate-limiter/redisRateLimiter.ts:
```diff
@@ -0,0 +1,108 @@
+    // ^ If a command does not return a reply within a set number of milliseconds,
+    // a "Command timed out" error will be thrown.
+  });
+  // TODO Should we set reconnectOnError to reconnect on specific errors?
```
yes if possible, thanks. This way we can reconnect for example on Redis restart or temporary connection lost

galales commented on packages/commons/src/rate-limiter/redisRateLimiter.ts:
```diff
@@ -0,0 +1,108 @@
+    // automatic failover disabled: https://github.com/redis/ioredis?tab=readme-ov-file#reconnect-on-error
+
+    // TODO figure out redis version we are using and set
+    // useRedisPackage accordingly: https://github.com/animir/node-rate-limiter-flexible/wiki/Redis#redis-package
```
we use Redis 7+

galales commented on packages/commons/src/rate-limiter/redisRateLimiter.ts:
```diff
@@ -0,0 +1,108 @@
+  // Seems to be supported but we need to set a burst points limit
+  // and also a burst rate interval.
+  // In Scala I see burstRequests = config.maxRequests * config.burstPercentage
+  // but what about the burst interval? Shoul we config.rateIntervalSeconds * config.burstPercentage?
```
this seems a good enough approximation, we can always tune this up in case of need

ecamellini commented:
@galales addressed all comments and added an automated test also to test the behavior of the "bursty" rate limiter as implemented by the library. Could you review the PR again and check the test file to see if it is the expected behavior?

Viktor-K reviewed (approved):
👍🏻 All looks good!

ecamellini commented:
@galales / @Viktor-K As suggested by Vittorio, I changed the PR to use `'redis'` instead of `ioredis` as the Redis client, since it is the one officially supported and maintained (even though the difference is confusing, being also `ioredis` in the Redis organization on GH. See: https://github.com/redis/node-redis/issues/2658#issuecomment-1887863434)
I think code also looks better...
